src/DGEList.py

```diff
@@ -98,16 +111,31 @@ def counts(self, counts: np.ndarray) -> np.matrix:
             counts: Columns correspond to samples and row to genes.
 
         """
+        if counts is None:
+            self._counts = None
+            return
+
+        if hasattr(self, '_counts'):
+            # do checks for things here.  You shouldn't modify counts
+            # if it has already been set.  Create a new obj.
+            if hasattr(self, '_samples') and self._samples is not None:
+                sample_count, gene_count = counts.shape
+                if sample_count != self._samples.shape[0] or gene_count != self._genes.shape[0]:
+                    raise ValueError("Attempting to substitute counts data "
+                                     "into DGEList object with different "
+                                     "dimensions fails.")
+
         if not isinstance(counts, np.ndarray):
             raise TypeError('Counts matrix must be of type ``np.ndarray``.')
-        elif np.any(counts < 0):
-            raise ValueError('Counts matrix cannot contain negative values.')
-        elif np.any(counts == np.nan):
+        if np.isnan(counts).any():
             raise ValueError('Counts matrix must have only real values.')
+        if (counts < 0).any():
+            raise ValueError('Counts matrix cannot contain negative values.')
         if self.to_remove_zeroes:
+            # this is not working.  Does not remove rows with only zeros.
             counts = counts[np.all(counts != 0, axis=1)]
 
-        self._counts = np.matrix(counts)
+        self._counts = counts
```

Comment:
You need to cast to at least `np.array()` here.

Reply:
Samples is already defined in the function header as an np.array or None.  I don't understand why you need to cast it.